src/Processors/Transforms/RollupTransform.cpp

```diff
@@ -1,36 +1,83 @@
 #include <Processors/Transforms/RollupTransform.h>
 #include <Processors/Transforms/TotalsHavingTransform.h>
 #include <Processors/QueryPlan/AggregatingStep.h>
+#include <Poco/Logger.h>
+#include "Common/logger_useful.h"
+#include "Columns/ColumnNullable.h"
 
 namespace DB
 {
 
-RollupTransform::RollupTransform(Block header, AggregatingTransformParamsPtr params_)
-    : IAccumulatingTransform(std::move(header), appendGroupingSetColumn(params_->getHeader()))
+GroupByModifierTransform::GroupByModifierTransform(Block header, AggregatingTransformParamsPtr params_, bool use_nulls_)
+    : IAccumulatingTransform(std::move(header), generateOutputHeader(params_->getHeader(), params_->params.keys, use_nulls_))
     , params(std::move(params_))
-    , aggregates_mask(getAggregatesMask(params->getHeader(), params->params.aggregates))
+    , use_nulls(use_nulls_)
 {
     keys.reserve(params->params.keys_size);
     for (const auto & key : params->params.keys)
         keys.emplace_back(input.getHeader().getPositionByName(key));
+
+    intermediate_header = getOutputPort().getHeader();
+    intermediate_header.erase(0);
```

Comment:
Please, explain why first column is removed.

Reply:
First column is `__grouping_set`. It's added exactly before returning generated chunk. `intermediate_heade` is used to correctly create columns with defaults for input blocks (they don't have `__grouping_set` column).